File: new/common/src/main/java/net/shiroha233/roadweaver/features/path/pathlogic/core/StructureRoadOffsetService.java
package net.shiroha233.roadweaver.features.path.pathlogic.core;

import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.level.Level;
import net.shiroha233.roadweaver.config.ConfigService;
import net.shiroha233.roadweaver.core.model.RoadSegmentPlacement;
import net.shiroha233.roadweaver.core.model.StructureInfo;
import net.shiroha233.roadweaver.search.StructurePredictor;

import java.util.ArrayList;
import java.util.List;

/**
 * 结构道路偏移服务
 */
public final class StructureRoadOffsetService {
    private StructureRoadOffsetService() {}

    private enum StructureCategory {
        VILLAGE,
        OTHER,
        UNKNOWN
    }

    private static final int MATCH_TOLERANCE_BLOCKS = 16;

    @Deprecated
    public static BlockPos adjustEndpoint(ServerLevel level, BlockPos endpoint, BlockPos otherEnd) {
        return endpoint;
    }

    public static List<RoadSegmentPlacement> trimPathNearStructure(
            ServerLevel level,
            List<RoadSegmentPlacement> segments,
            BlockPos rawStart,
            BlockPos rawEnd) {
        
        if (segments == null || segments.size() < 3) return segments;
        if (!Level.OVERWORLD.equals(level.dimension())) return segments;
        
        int startOffset = getOffsetBlocksForEndpoint(level, rawStart);
        int endOffset = getOffsetBlocksForEndpoint(level, rawEnd);
        
        if (startOffset <= 0 && endOffset <= 0) return segments;
        
        int n = segments.size();
        int trimStart = 0;
        int trimEnd = n;
        
        if (startOffset > 0) {
            long offsetSq = (long) startOffset * startOffset;
            for (int i = 0; i < n; i++) {
                BlockPos pos = segments.get(i).middlePos();
                long dx = (long) pos.getX() - rawStart.getX();
                long dz = (long) pos.getZ() - rawStart.getZ();
                long distSq = dx * dx + dz * dz;
                if (distSq >= offsetSq) {
                    trimStart = i;
                    break;
                }
            }
        }
        
        if (endOffset > 0) {
            long offsetSq = (long) endOffset * endOffset;
            for (int i = n - 1; i >= 0; i--) {
                BlockPos pos = segments.get(i).middlePos();
                long dx = (long) pos.getX() - rawEnd.getX();
                long dz = (long) pos.getZ() - rawEnd.getZ();
                long distSq = dx * dx + dz * dz;
                if (distSq >= offsetSq) {
                    trimEnd = i + 1;
                    break;
                }
            }
        }
        
        if (trimStart >= trimEnd || trimEnd - trimStart < 3) {
            int mid = n / 2;
            trimStart = Math.max(0, mid - 2);
            trimEnd = Math.min(n, mid + 3);
        }
        
        return new ArrayList<>(segments.subList(trimStart, trimEnd));
    }

    public static int getOffsetBlocksForEndpoint(ServerLevel level, BlockPos endpoint) {
        StructureCategory cat = detectCategory(level, endpoint);
        var cfg = ConfigService.get();
        return switch (cat) {
            case VILLAGE -> cfg.villageRoadOffset();
            case OTHER -> cfg.otherStructureRoadOffset();
            default -> 0;
        };
    }

    private static StructureCategory detectCategory(ServerLevel level, BlockPos endpoint) {
        if (!Level.OVERWORLD.equals(level.dimension())) return StructureCategory.UNKNOWN;

        int cx = endpoint.getX() >> 4;
        int cz = endpoint.getZ() >> 4;
        int searchRadius = 1;
        int tolerance = Math.max(MATCH_TOLERANCE_BLOCKS, ConfigService.get().aStarStep() + 8);
        long tol2 = (long) tolerance * (long) tolerance;

        List<StructureInfo> villageInfos = StructurePredictor.predictOverworldStructuresInRect(
                level,
                cx - searchRadius, cz - searchRadius,
                cx + searchRadius, cz + searchRadius,
                true,
                List.of("#minecraft:village"),
                List.of()
        );
        if (villageInfos != null && !villageInfos.isEmpty()) {
            for (StructureInfo info : villageInfos) {
                BlockPos p = info.pos();
                long dx = (long) p.getX() - endpoint.getX();
                long dz = (long) p.getZ() - endpoint.getZ();
                if (dx * dx + dz * dz <= tol2) {
                    return StructureCategory.VILLAGE;
                }
            }
        }

        var cfg = ConfigService.get();
        List<String> whitelist = cfg.structureWhitelist();
        if (whitelist != null && !whitelist.isEmpty()) {
            List<StructureInfo> otherInfos = StructurePredictor.predictOverworldStructuresInRect(
                    level,
                    cx - searchRadius, cz - searchRadius,
                    cx + searchRadius, cz + searchRadius,
                    true,
                    whitelist,
                    cfg.structureBlacklist()
            );
            if (otherInfos != null && !otherInfos.isEmpty()) {
                for (StructureInfo info : otherInfos) {
                    BlockPos p = info.pos();
                    long dx = (long) p.getX() - endpoint.getX();
                    long dz = (long) p.getZ() - endpoint.getZ();
                    if (dx * dx + dz * dz <= tol2) {
                        return StructureCategory.OTHER;
                    }
                }
            }
        }

        return StructureCategory.UNKNOWN;
    }
}

File: common/src/main/java/net/shiroha233/roadweaver/features/path/pathlogic/core/StructureRoadOffsetService.kt
package net.shiroha233.roadweaver.features.path.pathlogic.core

import net.minecraft.core.BlockPos
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.level.Level
import net.shiroha233.roadweaver.config.ConfigService
import net.shiroha233.roadweaver.helpers.Records
import net.shiroha233.roadweaver.search.StructurePredictor
import kotlin.math.max
import kotlin.math.min

/**
 * 专门负责根据结构类型决定道路端点需要从结构点向外缩进多少格。
 */
object StructureRoadOffsetService {
    private enum class StructureCategory {
        VILLAGE,
        OTHER,
        UNKNOWN
    }

    // 当坐标与预测结构点不完全重合时，允许的匹配容差（半径，单位：方块）
    private const val MATCH_TOLERANCE_BLOCKS = 16

    /**
     * @deprecated 使用 trimPathNearStructure 代替，在寻路完成后裁剪路径
     */
    @Deprecated("Use trimPathNearStructure")
    @JvmStatic
    fun adjustEndpoint(level: ServerLevel, endpoint: BlockPos, otherEnd: BlockPos): BlockPos {
        return endpoint
    }

    /**
     * 在 A* 寻路完成后，裁剪掉进入结构保护区的路段。
     */
    @JvmStatic
    fun trimPathNearStructure(
        level: ServerLevel,
        segments: List<Records.RoadSegmentPlacement>?,
        rawStart: BlockPos,
        rawEnd: BlockPos
    ): List<Records.RoadSegmentPlacement>? {
        if (segments == null || segments.size < 3) return segments
        if (Level.OVERWORLD != level.dimension()) return segments

        val startOffset = getOffsetBlocksForEndpoint(level, rawStart)
        val endOffset = getOffsetBlocksForEndpoint(level, rawEnd)

        if (startOffset <= 0 && endOffset <= 0) return segments

        val n = segments.size
        var trimStart = 0
        var trimEnd = n

        if (startOffset > 0) {
            val offsetSq = startOffset.toLong() * startOffset.toLong()
            for (i in 0 until n) {
                val pos = segments[i].middlePos
                val dx = pos.x.toLong() - rawStart.x.toLong()
                val dz = pos.z.toLong() - rawStart.z.toLong()
                val distSq = dx * dx + dz * dz
                if (distSq >= offsetSq) {
                    trimStart = i
                    break
                }
            }
        }

        if (endOffset > 0) {
            val offsetSq = endOffset.toLong() * endOffset.toLong()
            for (i in (n - 1) downTo 0) {
                val pos = segments[i].middlePos
                val dx = pos.x.toLong() - rawEnd.x.toLong()
                val dz = pos.z.toLong() - rawEnd.z.toLong()
                val distSq = dx * dx + dz * dz
                if (distSq >= offsetSq) {
                    trimEnd = i + 1
                    break
                }
            }
        }

        if (trimStart >= trimEnd || trimEnd - trimStart < 3) {
            val mid = n / 2
            trimStart = max(0, mid - 2)
            trimEnd = min(n, mid + 3)
        }

        return ArrayList(segments.subList(trimStart, trimEnd))
    }

    /**
     * 获取指定端点的保护区半径（方块数）
     */
    @JvmStatic
    fun getOffsetBlocksForEndpoint(level: ServerLevel, endpoint: BlockPos): Int {
        val cat = detectCategory(level, endpoint)
        val cfg = ConfigService.get()
        return when (cat) {
            StructureCategory.VILLAGE -> cfg.villageRoadOffset()
            StructureCategory.OTHER -> cfg.otherStructureRoadOffset()
            else -> 0
        }
    }

    private fun detectCategory(level: ServerLevel, endpoint: BlockPos): StructureCategory {
        if (Level.OVERWORLD != level.dimension()) return StructureCategory.UNKNOWN

        val cx = endpoint.x shr 4
        val cz = endpoint.z shr 4
        val searchRadius = 1

        val tolerance = max(MATCH_TOLERANCE_BLOCKS, ConfigService.get().aStarStep() + 8)
        val tol2 = tolerance.toLong() * tolerance.toLong()

        val villageInfos = StructurePredictor.predictOverworldStructuresInRect(
            level,
            cx - searchRadius,
            cz - searchRadius,
            cx + searchRadius,
            cz + searchRadius,
            true,
            listOf("#minecraft:village"),
            listOf()
        )
        if (!villageInfos.isNullOrEmpty()) {
            for (info in villageInfos) {
                val p = info.pos
                val dx = p.x.toLong() - endpoint.x.toLong()
                val dz = p.z.toLong() - endpoint.z.toLong()
                val d2 = dx * dx + dz * dz
                if (d2 <= tol2) {
                    return StructureCategory.VILLAGE
                }
            }
        }

        val cfg = ConfigService.get()
        val whitelist = cfg.structureWhitelist()
        if (!whitelist.isNullOrEmpty()) {
            val otherInfos = StructurePredictor.predictOverworldStructuresInRect(
                level,
                cx - searchRadius,
                cz - searchRadius,
                cx + searchRadius,
                cz + searchRadius,
                true,
                whitelist,
                cfg.structureBlacklist()
            )
            if (!otherInfos.isNullOrEmpty()) {
                for (info in otherInfos) {
                    val p = info.pos
                    val dx = p.x.toLong() - endpoint.x.toLong()
                    val dz = p.z.toLong() - endpoint.z.toLong()
                    val d2 = dx * dx + dz * dz
                    if (d2 <= tol2) {
                        return StructureCategory.OTHER
                    }
                }
            }
        }

        return StructureCategory.UNKNOWN
    }
}
